SimCalorimetry/HGCalAssociatorProducers/plugins/LayerClusterAssociatorByEnergyScoreImpl.cc

```diff
@@ -0,0 +1,531 @@
+  // 1. Extract collections and filter CaloParticles, if required
+  const auto& clusters = *cCCH.product();
+  const auto& caloParticles = *cPCH.product();
+  auto nLayerClusters = clusters.size();
+  //Consider CaloParticles coming from the hard scatterer, excluding the PU contribution.
+  auto nCaloParticles = caloParticles.size();
+  std::vector<size_t> cPIndices;
+  //Consider CaloParticles coming from the hard scatterer
+  //excluding the PU contribution and save the indices.
+  for (unsigned int cpId = 0; cpId < nCaloParticles; ++cpId) {
+    if (hardScatterOnly_ && (caloParticles[cpId].g4Tracks()[0].eventId().event() != 0 or
+                             caloParticles[cpId].g4Tracks()[0].eventId().bunchCrossing() != 0)) {
+      LogDebug("LayerClusterAssociatorByEnergyScoreImpl")
+          << "Excluding CaloParticles from event: " << caloParticles[cpId].g4Tracks()[0].eventId().event()
+          << " with BX: " << caloParticles[cpId].g4Tracks()[0].eventId().bunchCrossing() << std::endl;
+      continue;
+    }
+    cPIndices.emplace_back(cpId);
+  }
+  nCaloParticles = cPIndices.size();
+
+  // Initialize cPOnLayer. To be returned outside, since it contains the
+  // information to compute the CaloParticle-To-LayerCluster score.
+  hgcal::caloParticleToLayerCluster cPOnLayer;
+  cPOnLayer.resize(nCaloParticles);
+  for (unsigned int i = 0; i < nCaloParticles; ++i) {
+    cPOnLayer[i].resize(layers_ * 2);
+    for (unsigned int j = 0; j < layers_ * 2; ++j) {
+      cPOnLayer[i][j].caloParticleId = i;
+      cPOnLayer[i][j].energy = 0.f;
+      cPOnLayer[i][j].hits_and_fractions.clear();
+      //cPOnLayer[i][j].layerClusterIdToEnergyAndScore.reserve(nLayerClusters); // Not necessary but may improve performance
+    }
+  }
+
+  // Fill detIdToCaloParticleId_Map and update cPOnLayer
+  std::unordered_map<DetId, std::vector<hgcal::detIdInfoInCluster>> detIdToCaloParticleId_Map;
+  for (const auto& cpId : cPIndices) {
+    const SimClusterRefVector& simClusterRefVector = caloParticles[cpId].simClusters();
+    for (const auto& it_sc : simClusterRefVector) {
+      const SimCluster& simCluster = (*(it_sc));
+      const auto& hits_and_fractions = simCluster.hits_and_fractions();
+      for (const auto& it_haf : hits_and_fractions) {
+        DetId hitid = (it_haf.first);
+        int cpLayerId = recHitTools_->getLayerWithOffset(hitid) + layers_ * ((recHitTools_->zside(hitid) + 1) >> 1) - 1;
+        std::map<DetId, const HGCRecHit*>::const_iterator itcheck = hitMap_->find(hitid);
+        if (itcheck != hitMap_->end()) {
+          const HGCRecHit* hit = itcheck->second;
+          auto hit_find_it = detIdToCaloParticleId_Map.find(hitid);
+          if (hit_find_it == detIdToCaloParticleId_Map.end()) {
+            detIdToCaloParticleId_Map[hitid] = std::vector<hgcal::detIdInfoInCluster>();
+            detIdToCaloParticleId_Map[hitid].emplace_back(hgcal::detIdInfoInCluster{cpId, it_haf.second});
```

Comment:
you don't need to call the object constructor with emplace (that's the point of emplace)

Reply:
The [struct `detIdInfoInCluster`](https://github.com/cms-sw/cmssw/pull/29844/files#diff-5cc835ddfa8cd06584b636deddaf38bcR18) does not have a constructor so `emplace_back` does not work without it. Shall I define one?

Reply:
That would be preferable. (In fact, I would propose making all the structs into classes.)